DAOS-5371 tests: Junit results from Avocado not reported in Jenkins

phender opened the pull request:
When running functional tests from an DAOS RPM installation the non-root
user does not have permission to create files in the current directory.
This causes an IOError exception with the daos_core tests and the
test-specific *_results.xml is not created.

To resolve this issue the daos_core tests will now write the file into an
avocado test directory.

Skip-func-test: true
Skip-func-hw-test-small: true
Skip-func-hw-test-large: true

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

daosbuild1 commented on src/tests/ftest/util/daos_core_base.py:
```diff
@@ -21,6 +21,7 @@
   Any reproduction of computer software, computer software documentation, or
   portions thereof marked with this legend must also reproduce the markings.
 """
+import os
```
(pylint-unused-import) Unused import os

daosbuild1 replied:
(pylint-unused-import) Unused import os

daosbuild1 replied:
(pylint-unused-import) Unused import os

daosbuild1 replied:
(pylint-unused-import) Unused import os

daosbuild1 replied:
(pylint-unused-import) Unused import os

daosbuild1 replied:
(pylint-unused-import) Unused import os

daosbuild1 replied:
(pylint-unused-import) Unused import os

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/1/execution/node/59/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/1/execution/node/701/log

daosbuild1 commented on ftest.sh:
```diff
@@ -372,6 +372,15 @@ if grep \"testsuite.setAttribute('name', 'avocado')\" \
 /testsuite.setAttribute('name', 'avocado')/s/'avocado'/os.path.basename(os.path.dirname(result.logfile))/
 wq
 EOF
+# apply fix for https://github.com/avocado-framework/avocado/commit/bfab1c6b20f8489c343cfc8c8964d8905b0f72a1
```
(style)  line over 80 characters

daosbuild1 commented on ftest.sh:
```diff
@@ -372,6 +372,15 @@ if grep \"testsuite.setAttribute('name', 'avocado')\" \
+if grep \"from xml.dom.minidom import Document, Element\" \
+    /usr/lib/python2.7/site-packages/avocado/plugins/xunit.py; then
+    sudo ed <<EOF /usr/lib/python2.7/site-packages/avocado/plugins/xunit.py
+/from xml.dom.minidom import Document, Element/s/from xml.dom.minidom import Document/
```
(style)  line over 80 characters

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/2/execution/node/59/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/3/execution/node/365/log

daosbuild1 commented on ftest.sh:
```diff
@@ -373,6 +373,16 @@ if grep \"testsuite.setAttribute('name', 'avocado')\" \
 wq
 EOF
 fi
+# apply fix for https://github.com/avocado-framework/avocado/commit/bfab1c6b20f8489c343cfc8c8964d8905b0f72a1
```
(style)  line over 80 characters

daosbuild1 commented on ftest.sh:
```diff
@@ -373,6 +373,16 @@ if grep \"testsuite.setAttribute('name', 'avocado')\" \
+if grep \"from xml.dom.minidom import Document, Element\" \
+    /usr/lib/python2.7/site-packages/avocado/plugins/xunit.py; then
+    sudo ed <<EOF /usr/lib/python2.7/site-packages/avocado/plugins/xunit.py
+/from xml.dom.minidom import Document, Element/s/from xml.dom.minidom import Document/
```
(style)  line over 80 characters

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/4/execution/node/59/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/4/execution/node/701/log

daosbuild1 commented on ftest.sh:
```diff
@@ -373,6 +373,18 @@ if grep \"testsuite.setAttribute('name', 'avocado')\" \
 wq
 EOF
 fi
+# apply fix for https://github.com/avocado-framework/avocado/commit/bfab1c6b20f8489c343cfc8c8964d8905b0f72a1
```
(style)  line over 80 characters

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/5/execution/node/59/log

brianjmurrell commented on ftest.sh:
```diff
@@ -373,6 +373,18 @@ if grep \"testsuite.setAttribute('name', 'avocado')\" \
 wq
 EOF
 fi
+# apply fix for https://github.com/avocado-framework/avocado/commit/bfab1c6b20f8489c343cfc8c8964d8905b0f72a1
```
The commit message for that commit says it's Python 3 specific.  We are still using Python 2 for Avocado AFAIU.  Does this still apply to us?

phender replied:
Ah missed that.  As one would expect, since it's python 3 related, this 'fix' did not make a difference. I'll be removing it.

brianjmurrell commented on src/tests/ftest/util/daos_core_base.py:
```diff
@@ -82,11 +83,24 @@ def run_subtest(self):
+        Args:
+            result (CmdResult): result of the failed command.
+        """
+        filename = "".join([self.subtest_name, "_results.xml"])
```
Just curious why `join()` here instead of just `+`.

phender replied:
I tend to follow this recommendation https://www.python.org/dev/peps/pep-0008/#programming-recommendations

brianjmurrell replied:
Interesting recommendation.  Thanks for the pointer.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/6/execution/node/59/log

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/6/execution/node/672/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/7/execution/node/59/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/7/execution/node/701/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/8/execution/node/59/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3190/9/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/9/execution/node/804/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3190/11/testReport/(root)/

phender commented:
Failures seen in https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-3190/11:

- io/ior_small.py: https://jira.hpdd.intel.com/browse/DAOS-4989
- osa/osa_online_reintegration.py: https://jira.hpdd.intel.com/browse/DAOS-5416 (bug is resolved, but this run missed picking up the changes)

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/11/execution/node/770/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3190/12/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3190/12/testReport/(root)/

phender commented on src/tests/ftest/launch.py:
```diff
@@ -235,7 +235,8 @@ def set_test_environment(args):
 
-def get_output(cmd, check=True):
+
+def get_output(cmd, check=True, ):
```
Nope just a missed incomplete edit I forgot to remove when changing approaches

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3190/15/testReport/(root)/

phender commented:
Failures seen in https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-3190/15:

io/ior_small.py: https://jira.hpdd.intel.com/browse/DAOS-4989
osa/osa_online_reintegration.py: https://jira.hpdd.intel.com/browse/DAOS-5416

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3190/16/testReport/(root)/

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/16/execution/node/851/log

phender commented:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3190/16/#showFailuresLink includes the cmocka JUnit results.

phender commented:
> Philosophically, is there any point in continuing to run tests once the log collection disk is full?
> 
> This dovetails a bit on Amanda's patch where she and I even discussed what happens when the log collection disk fills, and stopping testing and generating a JUnit failure result if that does happen.
> 
> One could argue that even if the log collection disk fills (before the tests are finished running), if all of the tests actually pass, the test run should pass. But the counter-argument is that once that disk fills, the JUnit files no longer have space to be written, so we can't know if the remaining tests were successful or not as they are unreportable.
> 
> So while dealing with this seems like it belongs in Amanda's PR, the result of a filled log collection disk does cause the problem that this PR is aiming to solve -- namely that if the log collection disk does fill, JUnit files fail to be created, which means they fail to be reported. How is this prevented in this PR particularly given that we are no longer checking the success of things like the `mkdir` on line 940 of `launch.py`?

Right, trying to continue when we first detect out of disk space (via a failed mkdir) was something I was testing during this PR.  As long as we can get to https://github.com/daos-stack/pipeline-lib/blob/master/vars/functionalTestPost.groovy#L34 when we're out of disk space, we should have the results that have executed, so I should remove the `check=False` entries on lines 940 and 946.  It was my intention and I missed it.

brianjmurrell commented:
> Right, trying to continue when we first detect out of disk space (via a failed mkdir) was something I was testing during this PR. As long as we can get to https://github.com/daos-stack/pipeline-lib/blob/master/vars/functionalTestPost.groovy#L34 when we're out of disk space, we should have the results that have executed,

But be missing the results for tests that ran after the collection disk was full.  How does that turn into a failed test result given that Jenkins doesn't know that some results are simply missing?

> so I should remove the `check=False` entries on lines 940 and 946. It was my intention and I missed it.

Maybe when those fail, a new JUnit result file is generated to say that the disk filled so the run was aborted?

phender commented on ci/functional/job_cleanup.sh:
```diff
@@ -8,9 +8,10 @@ rm -rf install/lib/daos/TESTING/ftest/avocado/job-results/job-*/html/
-rm -rf "Functional/"
-mkdir "Functional/"
+if [ -e install/lib/daos/TESTING/ftest/avocado/job-results/latest ]; then
+  unlink install/lib/daos/TESTING/ftest/avocado/job-results/latest
```
Honestly, I'm not sure.  The '-e' test should detect if the file exists regardless of type (one could argue '-L' may make sense here, given the action taken if true).  Launch.py relies heavily on the existence of the 'latest' directory - without it we would not have the 'daos_logs' and 'daos_configs' directories and their files in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3190/16/artifact/Functional/job-2020-08-13T03.39-bde3265-daos_test-daos_core_test/.  So we have evidence that it existed and by virtue of it not existing in artifact/Functional we know it was removed (or someow otherwise not included in the [move](https://github.com/daos-stack/daos/blob/master/ci/functional/job_cleanup.sh#L24).

phender replied:
Ah, could it be https://github.com/daos-stack/daos/blob/master/ci/functional/job_cleanup.sh#L19-20?  Since it explicitly looks for "job-*" it would exclude "latest".

phender replied:
Still doesn't explain why the '-e' doesn't detect it though.

brianjmurrell replied:
Indeed.  It appears benign.  I just don't like not understanding why things that should be are not since the lack of understanding could be papering over other issues.

It doesn't help that `Jenkinsfile` size limitations have forced us into this popcorn-trail of indirections in following the flow of a test run any more.  :-(

phender replied:
I agree completely about not liking why the issue is not understood.  I am also questioning the need for the unlink command if we ignore the directory when moving the the test results to the 'Functional' directory.

phender replied:
I don't think the uncomfortableness outweighs the need to have other PRs correctly detect Functional HW Medium test failures.

brianjmurrell replied:
True.  But how do we avoid this just being forgotten and/or never addressed as it _looks_ benign?  Even if we open a ticket on it, it will never get any attention due to it's apparent benign-ness.  Until some other silent failure highlights it.  :-(

phender commented:
> But be missing the results for tests that ran after the collection disk was full. How does that turn into a failed test result given that Jenkins doesn't know that some results are simply missing?
>
It may or may not based upon what happened in the test, but won't the stage get marked as non-green because the ftest.sh exit status != 0?
 
> > so I should remove the `check=False` entries on lines 940 and 946. It was my intention and I missed it.
> 
> Maybe when those fail, a new JUnit result file is generated to say that the disk filled so the run was aborted?

How would we create a file when we're out of disk space?  We could pre-create the disk-full JUnit result before starting the test and then delete it if there is disk space available after test completion, but I suspect that could yield some false failures if something goes wrong in launch.py or ftest.sh.

phender commented:
Regardless of these good discussion points, shouldn't we address this in a follow on PR and close this one?  Do we need another run without the `check=False` entries in launch.py or can that be addressed later?

brianjmurrell commented:
> It may or may not based upon what happened in the test, but won't the stage get marked as non-green because the ftest.sh exit status != 0?

It should, if we exit launch.py` !0.  But we are really trying to prevent people from having to open up stage/step logs to figure out what went wrong.  We want all failures exposed in the Tests results tab.

> How would we create a file when we're out of disk space?

We might have to free up some space by removing the most recently copied log, as it's likely truncated anyway.

> We could pre-create the disk-full JUnit result before starting the test and then delete it if there is disk space available after test completion,

That's a good idea also.

> but I suspect that could yield some false failures if something goes wrong in launch.py or ftest.sh.

The pre-created file could reserve enough space in the file to include the results of the `df -h` so that it's assertion could be verified.

Or pre-create a large enough generic "run failed" JUnit file that is deleted when the run completes without issues.  At test end, the generic "something went wrong" JUnit file can be replaced with an ENOSPC file complete with `df -h` output if during cleanup it's determined that disk was out of space.

brianjmurrell commented:
> Regardless of these good discussion points, shouldn't we address this in a follow on PR and close this one?

I'm just thinking that ideally this PR addresses all of the issues known to cause missing test results, including the fallout from space issues.

> Do we need another run without the `check=False` entries in launch.py or can that be addressed later?

I think we have to meet the goal above, don't we?  Doesn't the `check=False` result in silent disk-full-no-junit-results situations?

phender commented:
> > Regardless of these good discussion points, shouldn't we address this in a follow on PR and close this one?
> 
> I'm just thinking that ideally this PR addresses all of the issues known to cause missing test results, including the fallout from space issues.
> 

Although handling no disk space issues is important (and some of the fixes are included in this PR), the additional work of reporting a stage failure (ftest.sh != 0) in the Test Results section and including more detail like `df` output - which I also agree is neccessary -  is out of the scope of the original problem.   I strongly believe that this should be addressed in a separate PR where:

- we can include a fake test to force the out of disk space scenario
- we can work through any implementation kinks while other PRs can at least get notifications of daos_core_test{,-rebuild} failures (by landing this PR as it stands)


> > Do we need another run without the `check=False` entries in launch.py or can that be addressed later?
> 
> I think we have to meet the goal above, don't we? Doesn't the `check=False` result in silent disk-full-no-junit-results situations?

I've removed the `check=False` code that was added during this PR as it should not be added.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/17/execution/node/823/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-3190/17/testReport/(root)/

phender commented:
Failures seen in https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-3190/17:

osa/osa_online_reintegration.py: https://jira.hpdd.intel.com/browse/DAOS-5416

The following errors are now being reported as part of this fix:

daos_test/daos_core_test.py: test_p: 
daos_test/daos_core_test.py: test_c: 
daos_test/daos_core_test-rebuild.py: *all*:

phender commented:
> Doesn't the:
> 
> ```
> Test-tag-hw-medium: pr,hw,medium,ib2,-daos_test_rebuild
> ```
> 
> make the test run actually avoid exactly the situation (one of the items) this PR is aiming to deal with, resulting in not testing that aspect of the changes?

It was a temporary tag to work around the disk-full issue to verify the other changes.  The latest commits use the normal tags.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3190/18/execution/node/910/log
